libraries/heft-config-file/src/ConfigurationFile.ts

```diff
@@ -668,6 +557,248 @@ export class ConfigurationFile<TConfigurationFile> {
+      originalValues: {} as TField
+    };
+    const result: TField = {
+      [CONFIGURATION_FILE_FIELD_ANNOTATION]: resultAnnotation
+    } as unknown as TField;
+
+    // Do a first pass to gather and strip the inheritance type annotations from the merging object
+    const currentObjectPropertyNames: Set<string> = new Set(Object.keys(currentObject || {}));
+    const filteredObjectPropertyNames: string[] = [];
+    const inheritanceTypeMap: Map<string, IPropertyInheritance<InheritanceType>> = new Map();
+    for (const propertyName of currentObjectPropertyNames) {
+      if (ignoreProperties && ignoreProperties.has(propertyName)) {
+        continue;
+      }
+      const inheritanceTypeMatches: RegExpMatchArray | null = propertyName.match(
+        CONFIGURATION_FILE_MERGE_BEHAVIOR_FIELD_REGEX
+      );
+      if (inheritanceTypeMatches && inheritanceTypeMatches.length === 2) {
+        const mergeTargetPropertyName: string = inheritanceTypeMatches[1];
+        const inheritanceTypeRaw: unknown | undefined = (currentObject || {})[propertyName];
+        if (!currentObjectPropertyNames.has(mergeTargetPropertyName)) {
+          throw new Error(
+            `Issue in processing configuration file property "${propertyName}". ` +
+              `An inheritance type was provided but no matching property was found`
+          );
+        } else if (typeof inheritanceTypeRaw !== 'string') {
+          throw new Error(
+            `Issue in processing configuration file property "${propertyName}". ` +
+              `An unsupported inheritance type was provided: ${JSON.stringify(inheritanceTypeRaw)}`
+          );
+        } else if (typeof (currentObject || {})[mergeTargetPropertyName] !== 'object') {
+          throw new Error(
+            `Issue in processing configuration file property "${propertyName}". ` +
+              `An inheritance type was provided for a property that is not an object`
+          );
+        }
+        switch (inheritanceTypeRaw.toLowerCase()) {
+          case 'append':
+            inheritanceTypeMap.set(mergeTargetPropertyName, { inheritanceType: InheritanceType.append });
+            break;
+          case 'merge':
+            inheritanceTypeMap.set(mergeTargetPropertyName, { inheritanceType: InheritanceType.merge });
+            break;
+          case 'replace':
+            inheritanceTypeMap.set(mergeTargetPropertyName, { inheritanceType: InheritanceType.replace });
+            break;
+          default:
+            throw new Error(
+              `Issue in processing configuration file property "${propertyName}". ` +
+                `An unsupported inheritance type was provided: "${inheritanceTypeRaw}"`
+            );
+        }
+      } else {
+        filteredObjectPropertyNames.push(propertyName);
+      }
+    }
+
+    // We only filter the currentObject because the parent object should already be filtered
+    const propertyNames: Set<string> = new Set<string>([
```

Comment:
Change these `string` types to `keyof TField` and you should be able to remove the `as any`s below. You'll need to cast some values to `keyof TField`.

If that doesn't work, cast to `Record<string, IPropertyInheritance<InheritanceType>>` instead below.